### DISCUSSION TITLE
Inline Query und Chatid

### DISCUSSION THREAD

**Discussion by averlon** [windkh/node-red-contrib-telegrambot#340] [General]
Ich bitte um Unterstützung um eine Inline Query hinzubekommen.

Meine Annahme:
Soweit ich die Doku verstanden habe, kann ein Benutzer in irgendeinem Chat (Channel oder Group)
mit "@botname command" eine Inline Query starten.

In NodeRed wird das mit einem Event-Node vom Event "Inline Query" als Eingang aufgefangen.

Ein Beispiel:
```
[
    {
        "id": "45960b7d9cc584fd",
        "type": "telegram event",
        "z": "33667947.7eff86",
        "name": "",
        "bot": "f37b455ea89fd0ca",
        "event": "inline_query",
        "autoanswer": true,
        "x": 110,
        "y": 1320,
        "wires": [
            [
                "5ba32f5326eedc17",
                "c5b80a01a8cd4427"
            ]
        ]
    },
    {
        "id": "5ba32f5326eedc17",
        "type": "debug",
        "z": "33667947.7eff86",
        "name": "av_tel inline_query",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "true",
        "targetType": "full",
        "statusVal": "",
        "statusType": "auto",
        "x": 310,
        "y": 1280,
        "wires": []
    },
    {
        "id": "c5b80a01a8cd4427",
        "type": "change",
        "z": "33667947.7eff86",
        "name": "inline_query",
        "rules": [
            {
                "t": "delete",
                "p": "payload",
                "pt": "msg"
            },
            {
                "t": "set",
                "p": "payload",
                "pt": "msg",
                "to": "{}",
                "tot": "json"
            },
            {
                "t": "set",
                "p": "payload.type",
                "pt": "msg",
                "to": "message",
                "tot": "str"
            },
            {
                "t": "set",
                "p": "payload.chatId",
                "pt": "msg",
                "to": "-1001509670149",
                "tot": "num"
            },
            {
                "t": "set",
                "p": "payload.content",
                "pt": "msg",
                "to": "Was ist gefragt?",
                "tot": "str"
            },
            {
                "t": "set",
                "p": "payload.options.reply_markup",
                "pt": "msg",
                "to": "$string(\t   {\t       \"resize_keyboard\":true,\t       \"one_time_keyboard\":true,\t       \"inline_keyboard\":[\t           [\t               {\t                   \"text\": \"Hilfe alle Kommandos\",\t                   \"callback_data\": \"hilfe\"                             \t               },\t               {\t                   \"text\": \"Abonnenten anzeigen\",\t                   \"callback_data\": \"abo\"                             \t               }                 \t           ]\t       ]    \t   } \t)",
                "tot": "jsonata"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 290,
        "y": 1320,
        "wires": [
            [
                "584fb6a9678bedb7",
                "5c4006e5b468f623"
            ]
        ]
    },
    {
        "id": "584fb6a9678bedb7",
        "type": "telegram sender",
        "z": "33667947.7eff86",
        "name": "",
        "bot": "f37b455ea89fd0ca",
        "haserroroutput": false,
        "outputs": 1,
        "x": 550,
        "y": 1320,
        "wires": [
            []
        ]
    },
    {
        "id": "f37b455ea89fd0ca",
        "type": "telegram bot",
        "botname": "averlon_bot",
        "usernames": "",
        "chatids": "",
        "baseapiurl": "",
        "updatemode": "polling",
        "pollinterval": "300",
        "usesocks": false,
        "sockshost": "",
        "socksprotocol": "socks5",
        "socksport": "6667",
        "socksusername": "anonymous",
        "sockspassword": "",
        "bothost": "",
        "botpath": "",
        "localbotport": "8443",
        "publicbotport": "8443",
        "privatekey": "",
        "certificate": "",
        "useselfsignedcertificate": false,
        "sslterminated": false,
        "verboselogging": false
    }
]
``` 

Der Debug vom Event zeigt:
```
{
  "payload": {
    "chatId": undefined,
    "type": "inline_query",
    "content": "",
    "inlineQueryId": "4213605307446235708",
    "offset": "",
    "from": {
      "id": 123456,
      "is_bot": false,
      "first_name": "Vorname",
      "last_name": "Nachname",
      "username": "username",
      "language_code": "de"
    }
  },
  "originalMessage": {
    "id": "4213605307446235708",
    "from": {
      "id": 123456,
      "is_bot": false,
      "first_name": "Vorname",
      "last_name": "Nachname",
      "username": "username",
      "language_code": "de"
    },
    "chat_type": "channel",
    "query": "",
    "offset": ""
  },
  "_msgid": "c1a686e07c652caf"
}
``` 

Jetzt ist die chatId des Channels oder der Group in dem die Inline Query abgesetzt wurde "undefinded".

Wenn jetzt die Antwortmöglichkeiten geschickt werden, wohin sollen die denn gehen? Die Antwortmöglichkeiten sollen doch da erscheinen, wo die Inline Query angefordert wurde!
An dieser Stelle bin ich etwas verloren!

Kann mir hier jemand aufs Pferd helfen?

Anmerkung:
Ich konnte erfolgreich eine Inline Query erzeugen und senden, wenn der Event ein Command war. In diesem Fall wird über den command-node die chatId mitgeliefert.

Danke

Zusatzfrage: Was bewirkt der Schalter "Auto-Answer"?

**Comment by windkh** [comment]
https://github.com/windkh/node-red-contrib-telegrambot/blob/master/examples/inlinequery.json
Hast du das Beispiel angeschaut?

**Comment by windkh** [comment]
Auto-Answer bezieht sich auf callback_query der mit answerCallbackQuery automatisch beantwortet werden kann. Alternativ kann man das im Flow selbst machen

**Comment by averlon** [comment]
Ich gebe zu, ich habe "inline_query", "inline_keyboard" und "custom_keyboard" etwas durcheinander geworfen.

Das Beispiel funktioniert!

Aber:

Ich wollte dann einen Button in die "inline_query" einfügen und habe dann die Doku der README gelesen, nachdem der erste Versuch nicht funktioniert hat.

Dort finde ich unter "Issueing API commands", in der Tabelle:

msg.payload.type = inline_query

Dann soll

msg.payload.content = { inlineQueryId (string), results (array of InlineQueryResult) }

Das würde für mich dann so aussehen:

msg.payload.content.inlineQueryId = "..."

msg.payload.content.results = []

Wenn ich das Beispiel aber nicht falsch interpretiere, dann wird dort das Array in msg.payload.results übergeben.

Ich bin kein super Entwickler. Ich muss mich weiter an Try&Error entlanghangeln und brauche manchmal eben etwas Hilfe.

**Comment by averlon** [comment]
Ich habe jetzt einiges getestet mit inline_query + button. Ergebnis:

Wird nur eine inline_query beantwortet, muss das in msg.payload.results stehen.

Wird zusätzlich ein Button gesendet, muss das in 

msg.payload.options.results und

msg.payload.options.button stehen!

mmh!

**Reply by windkh** [reply]
Verstehe ich nicht

**Comment by averlon** [comment]
schade - dann habe ich mich wohl nicht gut ausgedrückt.

**Comment by averlon** [comment]
So funktioniert eine inline_query:

```
[
    {
        "id": "6b04297c633d13e2",
        "type": "change",
        "z": "e69d8542b2668ba0",
        "d": true,
        "name": "inline_query",
        "rules": [
            {
                "t": "set",
                "p": "payload.results",
                "pt": "msg",
                "to": "[\t   {\t       \"type\" : \"article\",\t       \"id\" : \"1\",\t       \"title\" : \"Antwort 1\",\t       \"input_message_content\" : {\t           \"message_text\" : \"The message 1\",\t           \"parse_mode\" : \"HTML\",\t           \"disable_web_page_preview\" : false       \t       },\t       \"url\" : \"https://www.gwfwsfi.de\",\t       \"description\" : \"Beschreibung der ersten Antwort\"   \t   },\t   {\t       \"type\" : \"article\",\t       \"id\" : \"2\",\t       \"title\" : \"Antwort 2\",\t       \"input_message_content\" : {\t           \"message_text\" : \"The message 2\",\t           \"parse_mode\" : \"HTML\",\t           \"disable_web_page_preview\" : false       \t       }   \t   } \t]",
                "tot": "jsonata"
            },
            {
                "t": "set",
                "p": "payload.options",
                "pt": "msg",
                "to": "{}",
                "tot": "json"
            },
            {
                "t": "set",
                "p": "payload.options.cache_time",
                "pt": "msg",
                "to": "5",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 310,
        "y": 660,
        "wires": [
            [
                "14cc35cb61baab7e"
            ]
        ]
    }
]
```

Das entspricht auch dem Beispiel!

So funktioniert eine inline_query mit einem buttton:

```
[
    {
        "id": "0b2c423515e99d14",
        "type": "change",
        "z": "e69d8542b2668ba0",
        "d": true,
        "name": "inline_query+button",
        "rules": [
            {
                "t": "set",
                "p": "payload.options",
                "pt": "msg",
                "to": "{}",
                "tot": "json"
            },
            {
                "t": "set",
                "p": "payload.options.results",
                "pt": "msg",
                "to": "[\t   {\t       \"type\" : \"article\",\t       \"id\" : \"1\",\t       \"title\" : \"Antwort 1\",\t       \"input_message_content\" : {\t           \"message_text\" : \"The message 1\",\t           \"parse_mode\" : \"HTML\",\t           \"disable_web_page_preview\" : false       \t       },\t       \"description\" : \"Beschreibung der ersten Antwort\"   \t   },\t   {\t       \"type\" : \"article\",\t       \"id\" : \"2\",\t       \"title\" : \"Antwort 2\",\t       \"input_message_content\" : {\t           \"message_text\" : \"The message 2\",\t           \"parse_mode\" : \"HTML\",\t           \"disable_web_page_preview\" : false       \t       }   \t   } \t]",
                "tot": "jsonata"
            },
            {
                "t": "set",
                "p": "payload.options.button",
                "pt": "msg",
                "to": "$string(\t   {\t       \"text\" : \"Das kannst du mich fragen:\",\t       \"start_parameter\" : \"xxx\"\t   }\t)",
                "tot": "jsonata"
            },
            {
                "t": "set",
                "p": "payload.options.cache_time",
                "pt": "msg",
                "to": "5",
                "tot": "num"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 280,
        "y": 780,
        "wires": [
            []
        ]
    }
]
```

Man kann erkennen, dass bei der inline_query die werte für die Query in msg.payload.results stehen.

Bei inline_query + button stehen sie aber in msg.payload.options.results.

Darauf muss man erst mal kommen!!!!

Die Dokumentation sagt, falls ich das nicht falsch interpretiere:

### Issueing API commands
Additionally to sending content, the sender node can be used to issue commands direct to the API. In this case the `msg.payload` elements contain (see examples for further details):
- `msg.payload.type` - one of the commands listed below
- `msg.payload.content` - required command arguments
- `msg.payload.options` (optional) - additional command arguments


The `msg.payload.type` needs to be set to one of the following values:
| *msg.payload.type* | *msg.payload.content* | *msg.payload.options* | *see also* |

| **inline_query** | { inlineQueryId (string), results (array of InlineQueryResult) } | optional arguments | https://core.telegram.org/bots/api#answerinlinequery |

Also sollen die Werte für die inline_query eigentlich in msg.payload.content.results stehen - meine ich!

**Comment by windkh** [comment]
_(Entwurf von Claude, dem KI-Assistenten von Anthropic. Gepostet vom Maintainer.)_

@averlon — danke, dass du deine funktionierende Lösung hier geteilt hast! Genau so ist es richtig.

Zur Klärung, weil die README an dieser Stelle missverständlich war: Der `telegram sender` liest die Felder für `answerInlineQuery` **direkt aus `msg.payload`** (nicht aus `msg.payload.content`):

- `msg.payload.inlineQueryId` — die Query-ID
- `msg.payload.results` — das Array der `InlineQueryResult`-Objekte
- `msg.payload.options` — alle weiteren `answerInlineQuery`-Parameter, u. a. **`button`** (das frühere `switch_pm_text`/`switch_pm_parameter` ist jetzt das `button`-Objekt), `cache_time`, `is_personal`, `next_offset`

Deshalb landet der Button bei dir unter `msg.payload.options.button` und die Ergebnisse unter `msg.payload.results` — das ist korrekt und beabsichtigt.

Ich habe die README-Tabelle ("Issueing API commands") entsprechend präzisiert, damit klar ist, dass `inlineQueryId`/`results` Top-Level-Felder sind und `button` & Co. in `options` gehören.

Markiere das gerne als Antwort. Beispiel-Flow zum Vergleich: [examples/inlinequery.json](https://github.com/windkh/node-red-contrib-telegrambot/blob/master/examples/inlinequery.json)
